tsne/bh_sne_src/tsne.h

```diff
@@ -49,12 +49,23 @@
 #endif
 
 extern "C" {
-
+// stateless t-SNE
 void DLL_PUBLIC run(double* X, int N, int D, double* Y, int no_dims,
                     double perplexity, double theta, int rand_seed,
                     bool skip_random_init, double* init, bool use_init,
                     int max_iter = 1000, int stop_lying_iter = 250,
                     int mom_switch_iter = 250);
+// stateful t-SNE
+struct TSNE;
```

Comment:
So the user only actually reads output via `Y` which they passed in?  Cool.  Nice to not have any of those internal state details be part of the C API.

Reply:
Yeah, once I realized this I just decided to make the whole state object opaque.